### PR TITLE
linux: enable CONFIG_SND_HDA_CODEC_CS8409

### PR DESCRIPTION
**Pull request by paepckehh**
## Description of changes

Add linux kernel support for Cirrus Logic CS8409 as module.

This will unlock on-board Sound for some 2016-2019 Apple Intel Hardware Series (iMac, MacBookAir, MacBookPro) - tested on MacBookPro14,1

This will solve issues:
https://github.com/NixOS/nixos-hardware/issues/688
https://github.com/NixOS/nixpkgs/issues/313843

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 1]
```diff
@@ -511,6 +511,7 @@ let
       # Support configuring jack functions via fw mechanism at boot
       SND_HDA_PATCH_LOADER = yes;
       SND_HDA_CODEC_CA0132_DSP = whenOlder "5.7" yes; # Enable DSP firmware loading on Creative Soundblaster Z/Zx/ZxR/Recon
+      SND_HDA_CODEC_CS8409 = module; # Cirrus Logic HDA Bridge CS8409
```
Comment (Mic92) [inline pkgs/os-specific/linux/kernel/common-config.nix:514]: I am still waiting for the build to finish but so far it looks like at least some of the  hardened kernel likes this option:

```
┃    ┌─ ⚠ linux-config-5.4.277 failed with exit code 1 after ⏱ 25s in buildPhase
┃ ┌─ ⏸ linux-hardened-5.4.277
┃ │  ┌─ ⚠ linux-config-5.10.218 failed with exit code 1 after ⏱ 23s in buildPhase
┃ ├─ ⏸ linux-hardened-5.10.218
┃ │  ┌─ ⚠ linux-config-4.19.315 failed with exit code 1 after ⏱ 19s in buildPhase
┃ ├─ ⏸ linux-hardened-4.19.315
┃ │  ┌─ ✔ linux-config-6.9.7 ⏱ 41s
┃ ├─ ⏵ linux-libre-6.9.7 (buildPhase) ⏱ 1h1m18s
┃ │  ┌─ ✔ linux-config-6.9.7 ⏱ 41s
┃ ├─ ⏵ linux-lqx-6.9.7 (buildPhase) ⏱ 1h1m20s
┃ ├─ ⏵ linux-zen-6.9.7 (buildPhase) ⏱ 1h1m43s
┃ ├─ ⏵ linux-hardened-6.8.11 (buildPhase) ⏱ 1h2m1s
┃ ├─ ⏵ linux-xanmod-6.6.33 (buildPhase) ⏱ 1h2m1s
┃ ├─ ⏵ linux-xanmod-6.9.4 (buildPhase) ⏱ 1h2m1s
┃ ├─ ⏵ linux-hardened-6.9.2 (buildPhase) ⏱ 1h2m1s
┃ ├─ ⏵ linux-6.9.7 (buildPhase) ⏱ 1h2m1s
┃ ├─ ⏵ linux-libre-6.6.36 (buildPhase) ⏱ 1h2m1s
┃ ├─ ⏵ linux-6.6.36 (buildPhase) ⏱ 1h2m1s
┃ ├─ ⏵ linux-hardened-6.1.92 (buildPhase) ⏱ 1h2m2s
┃ ├─ ⏵ linux-hardened-6.6.32 (buildPhase) ⏱ 1h2m2s
```

```
CHOICE: 1-3?, ANSWER:
GOT:
GOT: *
GOT: * Kernel Testing and Coverage
GOT: *
GOT: *
GOT: * KUnit - Enable support for unit tests
GOT: *
QUESTION: KUnit - Enable support for unit tests, NAME: KUNIT, ALTS: N/m/y/?, ANSWER: n
GOT: n
QUESTION: Notifier error injection, NAME: NOTIFIER_ERROR_INJECTION, ALTS: N/m/y/?, ANSWER: n
GOT: n
QUESTION: Fault-injections of functions, NAME: FUNCTION_ERROR_INJECTION, ALTS: N/y/?, ANSWER:
GOT:
QUESTION: Fault-injection framework, NAME: FAULT_INJECTION, ALTS: N/y/?, ANSWER:
GOT:
QUESTION: Code coverage for fuzzing, NAME: KCOV, ALTS: N/y/?, ANSWER:
GOT:
GOT: *
GOT: * Runtime Testing
GOT: *
QUESTION: Runtime Testing, NAME: RUNTIME_TESTING_MENU, ALTS: N/y/?, ANSWER: n
GOT: n
QUESTION: Memtest, NAME: MEMTEST, ALTS: Y/n/?, ANSWER: y
GOT: y
QUESTION: Microsoft Hyper-V driver testing, NAME: HYPERV_TESTING, ALTS: N/y/?, ANSWER:
GOT:
GOT: #
GOT: # configuration written to .config
GOT: #
GOT: make: Leaving directory '/build/linux-5.10.218'
warning: unused option: CFI_CLANG
warning: unused option: CONFIG_LDISC_AUTOLOAD
warning: unused option: DEBUG_INFO_BTF
warning: unused option: INET_DIAG_DESTROY
warning: unused option: INET_MPTCP_DIAG
warning: unused option: INET_RAW_DIAG
warning: unused option: INET_TCP_DIAG
warning: unused option: INET_UDP_DIAG
warning: unused option: IOMMU_DEFAULT_DMA_LAZY
warning: unused option: IOMMU_DEFAULT_DMA_STRICT
warning: unused option: IO_STRICT_DEVMEM
warning: unused option: POWER_RESET_GPIO
warning: unused option: POWER_RESET_GPIO_RESTART
warning: unused option: SECURITY_WRITABLE_HOOKS
error: unused option: SND_HDA_CODEC_CS8409
warning: unused option: UBSAN_LOCAL_BOUNDS
```

You should be able to reproduce this by building:

```
nix-build -A linuxKernel.packages.linux_5_10_hardened.kernel
```

Comment (Mic92) [reply pkgs/os-specific/linux/kernel/common-config.nix:514]: So these are all new regressions:

```
Result of `nixpkgs-review pr 322968` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>linuxPackages_4_19_hardened.kernel</li>
    <li>linuxPackages_4_19_hardened.kernel.dev</li>
    <li>linuxPackages_5_10_hardened.kernel</li>
    <li>linuxPackages_5_10_hardened.kernel.dev</li>
    <li>linuxPackages_5_4_hardened.kernel</li>
    <li>linuxPackages_5_4_hardened.kernel.dev</li>
  </ul>
</details>
<details>
  <summary>26 packages built:</summary>
  <ul>
    <li>linuxPackages-libre.kernel</li>
    <li>linuxPackages-libre.kernel.dev</li>
    <li>linuxPackages.kernel</li>
    <li>linuxPackages.kernel.dev</li>
    <li>linuxPackages_5_15_hardened.kernel</li>
    <li>linuxPackages_5_15_hardened.kernel.dev</li>
    <li>linuxPackages_6_1_hardened.kernel</li>
    <li>linuxPackages_6_1_hardened.kernel.dev</li>
    <li>linuxPackages_hardened.kernel (linuxPackages_6_6_hardened.kernel)</li>
    <li>linuxPackages_hardened.kernel.dev (linuxPackages_6_6_hardened.kernel.dev)</li>
    <li>linuxPackages_6_8_hardened.kernel</li>
    <li>linuxPackages_6_8_hardened.kernel.dev</li>
    <li>linuxPackages_6_9_hardened.kernel</li>
    <li>linuxPackages_6_9_hardened.kernel.dev</li>
    <li>linuxPackages_latest-libre.kernel</li>
    <li>linuxPackages_latest-libre.kernel.dev</li>
    <li>linuxPackages_latest.kernel</li>
    <li>linuxPackages_latest.kernel.dev</li>
    <li>linuxPackages_lqx.kernel</li>
    <li>linuxPackages_lqx.kernel.dev</li>
    <li>linuxPackages_xanmod.kernel</li>
    <li>linuxPackages_xanmod.kernel.dev</li>
    <li>linuxPackages_xanmod_latest.kernel (linuxPackages_xanmod_stable.kernel)</li>
    <li>linuxPackages_xanmod_latest.kernel.dev (linuxPackages_xanmod_stable.kernel.dev)</li>
    <li>linuxPackages_zen.kernel</li>
    <li>linuxPackages_zen.kernel.dev</li>
  </ul>
</details>
```

Comment (Mic92) [reply pkgs/os-specific/linux/kernel/common-config.nix:514]: These needs to be addressed before we can merge this.

Comment (paepckehh) [reply pkgs/os-specific/linux/kernel/common-config.nix:514]: Can we skip the backport part? The kernel option is less than a year old ... 

https://www.kernelconfig.io/config_snd_hda_codec_cs8409

Comment (Mic92) [reply pkgs/os-specific/linux/kernel/common-config.nix:514]: No backport required but we only should set this options for kernel that have support for it.

Comment (Mic92) [reply pkgs/os-specific/linux/kernel/common-config.nix:514]: Just set an appropriate `versionAtLeast` constraint.

Comment (paepckehh) [reply pkgs/os-specific/linux/kernel/common-config.nix]: kernel >=6.6 should be an safe option for official 24.05 supporting this hardware

**Comment by nate-moo** [comment]
Hello, I'm not sure if this is the right place to ask about this but I have been looking at the macbook pro 14,1 hardware modules, and saw this. It looks like its been merged into master and the cs8409 module is being loaded, but I still don't have sound output beyond an inactive and non-functioning card in the sound panel. Was I just hallucinating the merge to master or is there something possibly wrong on my end? Thanks for your time, Nate

**Comment by Mic92** [comment]
Sorry. I don't have this hardware for testing.

**Comment by snoweuph** [comment]
In on a macbook Pro 14,2 and was using fedora just a day ago and the upstream module worked under fedora. with nixos, on the otherhand I dont get any audio

**Comment by paepckehh** [comment]
please share (a link) to a dmesg and lspci -kvvv somewhere, apple seems to use a wild range of vendor/device ids here

**Comment by snoweuph** [comment]
Gonna do it as soon as im home later today!

**Comment by snoweuph** [comment]
`dmesg`: https://sourceb.in/trIYnGDydt
`lspci -kvvv` part1: https://sourceb.in/0IbTtroLw2
`lspci -kvvv` part2: https://sourceb.in/P2NDF6GfrD

**Comment by snoweuph** [comment]
@paepckehh does this help, or do you require and other information?

**Pull request merged**